File: epi_judge_python_solutions/evaluate_rpn.py
from typing import List

from test_framework import generic_test


def evaluate(expression: str) -> int:

    intermediate_results: List[int] = []
    delimiter = ','
    operators = {
        '+': lambda y, x: x + y,
        '-': lambda y, x: x - y,
        '*': lambda y, x: x * y,
        '/': lambda y, x: x // y
    }

    for token in expression.split(delimiter):
        if token in operators:
            intermediate_results.append(operators[token](
                intermediate_results.pop(), intermediate_results.pop()))
        else:  # token is a number.
            intermediate_results.append(int(token))
    return intermediate_results[-1]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('evaluate_rpn.py', 'evaluate_rpn.tsv',
                                       evaluate))

File: epi_judge_java_solutions/epi/EvaluateRpn.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.function.ToIntBiFunction;

public class EvaluateRpn {
  @EpiTest(testDataFile = "evaluate_rpn.tsv")

  public static int eval(String expression) {

    Deque<Integer> intermediateResults = new ArrayDeque<>();
    final String DELIMITER = ",";
    final Map<String, ToIntBiFunction<Integer, Integer>> OPERATORS = Map.of(
        "+",
        (y, x)
            -> x + y,
        "-", (y, x) -> x - y, "*", (y, x) -> x * y, "/", (y, x) -> x / y);
    for (String token : expression.split(DELIMITER)) {
      if (OPERATORS.get(token) != null) {
        intermediateResults.addFirst(
            OPERATORS.get(token).applyAsInt(intermediateResults.removeFirst(),
                                            intermediateResults.removeFirst()));
      } else { // token is a number.
        intermediateResults.addFirst(Integer.parseInt(token));
      }
    }
    return intermediateResults.removeFirst();
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "EvaluateRpn.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
